Add schema1 conversion on pull

dmcgowan opened the pull request:
This is a bit of madness but it is isolated to the remote docker package. This is enforcing that only schema2 gets uploaded into the containerd content store. Repeated pulls to the same image will result in re-creation of the content. Once we get labels on the content store we can probably use diff id labels to prevent this doing re-work to calculate the rootfs.

closes #851

dmcgowan commented:
@Random-Liu I tried this out on a few gcr.io images, can you test on some of the images you had in mind

Random-Liu commented:
@dmcgowan Thanks a lot! Will try this out this.

Random-Liu commented:
Several different problems here:
```
root@workstation:~# dist pull gcr.io/google_containers/pause:3.0
gcr.io/google_containers/pause:3.0:                                            resolved       |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4: done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:f112334343777b75be77ec1f835e3bbbe7d7bd46e27b6a2ae35c6b3cfea0987c: done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 0.4 s                                                                 total:  32.0 B (79.0 B/s)                                        
unpacking sha256:94fcf736c22932c8245837be07e084e5171b8a94db7234766bcda7e8152b2305...done

root@workstation:~# dist pull gcr.io/google_containers/pause:3.0
gcr.io/google_containers/pause:3.0: resolving      |--------------------------------------| 
elapsed: 0.1 s                      total:   0.0 B (0.0 B/s)                                         
dist: rpc error: code = Unknown desc = failed reading status of resume write: stat /var/lib/containerd/content/ingest/77403316d19e53ab944f93d0ec6e1f6ca4b306d3544635eb419e0fc6ad41482a/data: no such file or directory

root@workstation:~# dist pull gcr.io/google-containers/stress:v1
gcr.io/google-containers/stress:v1:                                            resolved       |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:2ce2382a5646300744a450bcd4d7c0e279fa400c5133044e5398a6fe81b5f554: done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 0.7 s                                                                 total:  1.0 Mi (1.5 MiB/s)                                       
unpacking sha256:93a3ab056c8d0499e56a434122d268ad8f57d015d49598c801077c49b7fc2e4e...done
dist: mismatched image rootfs and manifest layers

root@workstation:~# dist pull gcr.io/google_containers/busybox:1.24
gcr.io/google_containers/busybox:1.24:                                         resolved       |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:eeee0535bf3cec7a24bff2c6e97481afa3d37e2cdeff277c57cb5cbdb2fa9e92: done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4: exists         |++++++++++++++++++++++++++++++++++++++| 
elapsed: 0.5 s                                                                 total:  416.0  (831.8 KiB/s)                                     
unpacking sha256:6d40c361f005073d7d72837223b8785ce8153424167310b7d9ec12f32b3c682b...INFO[0000] Apply failure, attempting cleanup             error="wrong diff id calculated on extraction \"sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef\"" key="extract sha256:2bfa4e68beed35267dcda5e4c3018b4d6580111037a8aa9eca57dd201676a446"
done
dist: wrong diff id calculated on extraction "sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef"
```

dmcgowan commented:
@Random-Liu thanks for the additional test cases. I will take a look at some of those images. Looks like something needs to be cleaned up with the content store reader as well.

dmcgowan commented:
@Random-Liu there appears to be some issues with pull related to handling duplicated hashes in the same manifest. For schema2 we don't get this condition since empty layers do not produce layers in the manifest. I will get this fix and add better support for these empty layers.

dmcgowan commented:
Pushed a commit which should fix the v1 handling of empty layers. These images pre-date the inclusion of the `ThrowAway` field in the v1 compability. When that field is missing, I am just using the size field (or lack of having either) to determine whether the layer is empty. Getting rid of these empty layers makes the pull not encounter the lock issue and I suspect was also causing the diff id calculation to fail. That being said, there is still a race/concurrency bug related to handling of the layers. I have tracked down why that occurs but not sure what the fix is. That will be a separate PR to fix that.

Random-Liu commented:
@dmcgowan Yeah, just tried, and `pause` image seems to work. But others still have problems.

Mostly caused by this https://github.com/containerd/containerd/blob/master/cmd/dist/common.go#L200.

dmcgowan commented:
Let me try with a few more images, I was not able to get that mismatch error to show up.

Random-Liu commented:
> Let me try with a few more images, I was not able to get that mismatch error to show up.

I hit it with `gcr.io/google_containers/busybox:1.24` and `gcr.io/google_containers/nginx-slim:0.7`.

dmcgowan commented:
Pushed another update, there was one other place that wasn't properly handling the empty layer leading to a difference in length between the rootfs diffids and the manifest layer ids.

Random-Liu commented:
@dmcgowan Thanks! Tried 4 different images, most of them work like a charm!! Except:
```
# dist pull gcr.io/google-containers/stress:v1
gcr.io/google-containers/stress:v1: resolving      |--------------------------------------| 
elapsed: 0.2 s                      total:   0.0 B (0.0 B/s)                                         
dist: rpc error: code = Unknown desc = failed reading status of resume write: stat /var/lib/containerd/content/ingest/77403316d19e53ab944f93d0ec6e1f6ca4b306d3544635eb419e0fc6ad41482a/data: no such file or directory
```

BTW, does the race you mentioned in https://github.com/containerd/containerd/pull/968#issuecomment-306934339 still exist?

dmcgowan commented:
@Random-Liu great! Yes, I am working on the race right now. I have identified the bug and working on the fix.

Random-Liu commented:
@dmcgowan Hi, how is the status of this PR now? Is the bug fixed? :)

dmcgowan commented:
@Random-Liu fixed the contention bug. I am making one more change to this PR to better handle the empty layers. The logic I added is not true in all cases so going to just base the size off what is calculated and skip duplicates.

crosbymichael commented:
LGTM

dmcgowan commented:
@crosbymichael marked as WIP until I fix this last issue with empty layers

crosbymichael commented:
@dmcgowan fine, i'll take it back

Random-Liu commented:
@dmcgowan Thanks a lot for the update!

dmcgowan commented:
OK! Rebased and pushed update. I added a fix which handles the ambiguity around determining whether the layer is empty. In the process, I have really found out how bad schema1 is and our attempt to make it better by adding a `throwaway` field for empty layers actually made it worse! When the new field was added it was added with omitempty and also got rid of the `Size` field at the same time. What this means is that if a layer does not have a size or empty flag, then we cannot determine whether the layer is empty from the history, since older images with empty layers would also discard the `Size` parameter with omitempty. Anyway, please try it out and review.

Random-Liu commented:
@dmcgowan Works like a charm! I've tried 5 different schema 1 images, all work now! Thanks a lot!

crosbymichael commented:
LGTM
